internal/bull/upload.go

```diff
@@ -0,0 +1,67 @@
+package bull
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+type uploadResponse struct {
+	File          string `json:"file"`          // Original filename
+	SavedFullPath string `json:"savedFullPath"` // Full path to the saved file
+	SavedFilename string `json:"savedFilename"` // Filename as saved
+}
+
+// upload Saves a file to <page>-<timestamp>-<original filename>
+func (b *bullServer) upload(w http.ResponseWriter, r *http.Request) error {
+	if b.editor == "" {
+		return httpError(http.StatusForbidden, fmt.Errorf("running in read-only mode (-editor= flag)"))
+	}
+
+	// Limit the size of the uploaded file to 25MB
+	err := r.ParseMultipartForm(25 << 20)
+	if err != nil {
+		return err
+	}
+
+	file, handler, err := r.FormFile("file")
+	if err != nil {
+		log.Printf("error reading file: %v", err)
+		w.WriteHeader(http.StatusInternalServerError)
```

Comment:
I don’t think you need to explicitly mark the response as an internal server error: When you return a non-nil error, that should already result in an internal server error.